python/ray/ml/checkpoint.py

```diff
@@ -1,17 +1,90 @@
 import abc
-from typing import Dict, Optional
+import json
+import shutil
+import tarfile
+import tempfile
+
+import cloudpickle as pickle
+import os
+from typing import Any, Type, Optional, Dict, List
 
 import ray
+from ray.util.ml_utils.cloud import (
+    upload_to_bucket,
+    is_cloud_target,
+    download_from_bucket,
+)
+
+
+def write_metadata(path: str, metadata: Dict[str, Any]) -> None:
+    metadata_file = os.path.join(path, ".checkpoint_metadata")
+    with open(metadata_file, "wt") as fp:
+        json.dump(metadata, fp)
+
+
+def load_metadata(path: str) -> Dict[str, Any]:
+    metadata_file = os.path.join(path, ".checkpoint_metadata")
+    if os.path.exists(metadata_file):
+        with open(metadata_file, "rt") as fp:
+            return json.load(fp)
+
+
+def temporary_checkpoint_dir() -> str:
+    return tempfile.mkdtemp(prefix="checkpoint_tmp_", dir=os.getcwd())
+
+
+def _pack(path: str) -> bytes:
+    _, tmpfile = tempfile.mkstemp()
+    with tarfile.open(tmpfile, "w:gz") as tar:
+        tar.add(path, arcname="")
+
+    with open(tmpfile, "rb") as f:
+        stream = f.read()
+
+    os.remove(tmpfile)
+    return stream
+
+
+def _unpack(stream: bytes, path: str) -> str:
+    _, tmpfile = tempfile.mkstemp()
+
+    with open(tmpfile, "wb") as f:
+        f.write(stream)
+
+    with tarfile.open(tmpfile) as tar:
+        tar.extractall(path)
+
+    os.remove(tmpfile)
+    return path
 
 
 class Checkpoint(abc.ABC):
-    """Checkpoint interface"""
+    """Checkpoint interface.
+
+    This implementation provides interfaces to translate between
+    different checkpoint storage locations: Local FS storage, remote
+    node FS storage, data object, and cloud storage location.
+
+    The following metadata keys are introduced for correct serialization
```

Comment:
Does Tune need to necessarily inspect the storage details of the checkpoint?

For example, suppose Tune wants to move a checkpoint to the driver. Tune can then call `checkpoint.to_object_ref()` on the worker, and `checkpoint.from_object_ref()` on the driver, without actually needing to inspect the type of the checkpoint.

Reply:
Tune needs to know how a checkpoint was originally stored (e.g. as dict or directory) so it can provide it to the trainable in the same format.
Similarly, Tune cleans up checkpoints on several locations (e.g. deletes it from cloud storage or a remote node), so we need to know if a checkpoint exists on these locations as well.
Let's chat about this later